### PR TITLE
Fix Custom Text Preference Resolver caching issue

### PR DESCRIPTION
**Pull request by UdeshAthukorala**
## Purpose

- Resolves: https://github.com/wso2/product-is/issues/18781

### CODE REVIEW

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java [hunk 1]
```diff
@@ -331,6 +332,54 @@ Get ancestor organization ids (including itself) of a given organization. The li
...
+                            + currentTenantDomain, e);
+                }
+                return;
+            }
+        }
+
+        List<String> childOrganizationIds = new ArrayList<>();
+        childOrganizationIds.add(organizationId);
+
+        // Clear branding resolver caches by looping (breadth-first) through child organization hierarchy.
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java:358]:
```suggestion
        // Clear custom text resolver caches by looping (breadth-first) through child organization hierarchy.
```

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/constant/BrandingPreferenceMgtConstants.java [hunk 2]
```diff
@@ -92,7 +92,9 @@ public enum ErrorMessages {
         ERROR_CODE_ERROR_BUILDING_CUSTOM_TEXT_PREFERENCE("BRANDINGM_00029",
                 "Unable to build custom text preference from custom text configurations for tenant: %s."),
         ERROR_CODE_ERROR_BULK_DELETING_CUSTOM_TEXT_PREFERENCES("BRANDINGM_00030",
-                "Unable to bulk delete custom text preferences for tenant: %s.");
+                "Unable to bulk delete custom text preferences for tenant: %s."),
+        ERROR_CODE_ERROR_CLEARING_CUSTOM_TEXT_PREFERENCE_RESOLVER_CACHE_HIERARCHY("BRANDINGM_00031",
+                "Error while clearing branding preference resolver cache hierarchy for tenant: %s.");
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/constant/BrandingPreferenceMgtConstants.java:97]:
```suggestion
                "Error while clearing custom text preference resolver cache hierarchy for tenant: %s.");
```

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/UIBrandingPreferenceResolver.java [hunk 3]
```diff
@@ -48,9 +48,33 @@ public interface UIBrandingPreferenceResolver {
...
+     * the organization and all its children down the tree.
+     *
+     * @param currentTenantDomain   Tenant domain where the cache needs to be cleared.
+     * @param screen                Screen param of the custom text.
+     * @param locale                Locale param of the custom text.
+     * @throws BrandingPreferenceMgtException if any error occurred.
+     */
+    default void clearCustomTextResolverCacheHierarchy(String currentTenantDomain, String screen, String locale)
+            throws BrandingPreferenceMgtException {
+
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/UIBrandingPreferenceResolver.java:78]: If we are adding this as a default method shall we throw the `NotImplementedException` as in the `resolveCustomText` method?

File: components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java (hunk 1 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/24/commits/c3cef6944aa26bd9357f85464d19749758380ca7

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/constant/BrandingPreferenceMgtConstants.java (hunk 2 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/constant/BrandingPreferenceMgtConstants.java]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/24/commits/c3cef6944aa26bd9357f85464d19749758380ca7

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/UIBrandingPreferenceResolver.java (hunk 3 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/UIBrandingPreferenceResolver.java:78]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/24/commits/c3cef6944aa26bd9357f85464d19749758380ca7

**Review by dhaura — APPROVED** [review]

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**